### PR TITLE
gitActions: Add workflow to bump kubevirtci

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
This PR introduces a gitActions that will run the kubevirtci bumper every 3 months, in order to keep CNAO up to date.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RamLavi** [comment]
@oshoval wdyt?

**Comment by oshoval** [comment]
we might want to make sure 
https://github.com/kubevirt/cluster-network-addons-operator/pull/2131/files#diff-030affb1cecb4332ce886f9c3e14ed4740eff8868bfa708e5ae8f83f0bb133aaR4

this change is robust, i just changed it to make it work, not sure if it is consistent 
if it isnt it meant the job will fail and we dont see it

**Comment by RamLavi** [comment]
> Nice thanks lets change to 3M?

np

> 
> One thing to consider please, is the k8s version we use, sometimes the old one is dropped we can introduce auto detection of the latest version, wdyt? otoh sometimes kubevirtci has a newer version then the one we aim to, (and sometimes we use older one than the one we aim to) the best will be if we could query to detect the desired current k8s version automatically

Do you mean the KUBEVIRT_PROVIDER?

**Comment by oshoval** [comment]
> Do you mean the KUBEVIRT_PROVIDER?

yep

**Comment by RamLavi** [comment]
> > Do you mean the KUBEVIRT_PROVIDER?
> 
> yep

We can add another workflow to bump to the maximum KUBEVIRT_PROVIDER (with format of k8s-x.y). wdyt?

**Comment by RamLavi** [comment]
> we might want to make sure https://github.com/kubevirt/cluster-network-addons-operator/pull/2131/files#diff-030affb1cecb4332ce886f9c3e14ed4740eff8868bfa708e5ae8f83f0bb133aaR4
> 
> this change is robust, i just changed it to make it work, not sure if it is consistent if it isnt it meant the job will fail and we wont see it

If we see we get gitAction fails, we'll tackle it then I think..

**Comment by oshoval** [comment]
> We can add another workflow to bump to the maximum KUBEVIRT_PROVIDER (with format of k8s-x.y). wdyt?

it is problematic, because sometimes latest kubevirtci is more advance than the one we are working on, since the upstream CI team adds it before its needed

for sure we can discuss it on follow ups with Brian and Daniel

thanks

File: .github/workflows/kubevirtci-tag-bumper.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      - name: Extract new tag from cluster.sh
+        if: steps.changes.outputs.changed == 'true'
+        id: extract
+        run: |
+          NEW_TAG=$(grep -oP 'KUBEVIRTCI_TAG:-\K[^}]+' cluster/cluster.sh)
+          echo "Detected new tag: $NEW_TAG"
+          echo "tag=$NEW_TAG" >> $GITHUB_OUTPUT
+
+      - name: Create Pull Request
+        if: steps.changes.outputs.changed == 'true'
```
Comment (oshoval) [inline .github/workflows/kubevirtci-tag-bumper.yaml:46]: can we compact those steps to only one ?
maybe directly `if git diff ...; do stuff..` ?

**Comment by ashokpariya0** [comment]
Is it possible to check whether a tag supports multi-architecture? For example, the latest tag (2504011456-29ba6f0e) from [KubeVirtCI GitHub tags](https://github.com/kubevirt/kubevirtci/tags) does not support multi-architecture (you can verify this by checking the corresponding Go-CLI tag on [Quay.io](https://quay.io/repository/kubevirtci/gocli?tab=tags&tag=latest)), while the previous tag (2503281142-a291de27) does. Therefore, always using the latest tag might not be reliable

**Comment by oshoval** [comment]
> Is it possible to check whether a tag supports multi-architecture? For example, the latest tag (2504011456-29ba6f0e) from [KubeVirtCI GitHub tags](https://github.com/kubevirt/kubevirtci/tags) does not support multi-architecture (you can verify this by checking the corresponding Go-CLI tag on [Quay.io](https://quay.io/repository/kubevirtci/gocli?tab=tags&tag=latest)), while the previous tag (2503281142-a291de27) does. Therefore, always using the latest tag might not be reliable

either by podman inspect, or on quay.io web page it shows
(also with docker there is a trick, but with podman its easier)

**Comment by RamLavi** [comment]
> > Is it possible to check whether a tag supports multi-architecture? For example, the latest tag (2504011456-29ba6f0e) from [KubeVirtCI GitHub tags](https://github.com/kubevirt/kubevirtci/tags) does not support multi-architecture (you can verify this by checking the corresponding Go-CLI tag on [Quay.io](https://quay.io/repository/kubevirtci/gocli?tab=tags&tag=latest)), while the previous tag (2503281142-a291de27) does. Therefore, always using the latest tag might not be reliable
> 
> either by podman inspect, or on quay.io web page it shows (also with docker there is a trick, but with podman its easier)
> 
> we dont need to add such validation to git actions otoh for that we have the lanes you and us create, if it fails we will see there there was a little hiccup thats all, it doesnt mean we need to be extra safe from now on in this aspect

I agree with @oshoval . If any, a check like this should be inside the bump-kubevirt.sh script itself, and not in this workflow.

**Comment by oshoval** [comment]
> I agree with @oshoval . If any, a check like this should be inside the bump-kubevirt.sh script itself, and not in this workflow.

I would say no check at all, the lanes are already good enough
we can't add validation based on every problem we see, it is expansive to run and maintain
but thanks a lot for caring and contributing

File: .github/workflows/kubevirtci-tag-bumper.yaml (hunk 1 above)
Comment (RamLavi) [reply .github/workflows/kubevirtci-tag-bumper.yaml:46]: no, because this step (by peter-evans) is a standalone gitActions building block. If we wanted to bulk it together, we'd need to implement it ourselves like in the auto bumper case. But that's not necessary.

File: .github/workflows/kubevirtci-tag-bumper.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+          git config user.name "github-actions[bot]"
+          git config user.email "github-actions[bot]@users.noreply.github.com"
+
+      - name: Run bump-kubevirtci.sh
+        run: |
+          ./hack/bump-kubevirtci.sh
+
+      - name: Check for changes
+        id: changes
+        run: |
```
Comment (oshoval) [inline .github/workflows/kubevirtci-tag-bumper.yaml:30]: please consider 
`run: echo "changed=$(! git diff --quiet cluster/cluster.sh && echo true || echo false)" >> "$GITHUB_OUTPUT"`

File: .github/workflows/kubevirtci-tag-bumper.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
+name: Bump KubevirtCI Tag
+
+on:
+  schedule:
+    - cron: '0 0 1 1,5,9 *'  # 00:00 on the 1st of Jan, May, Sep
```
Comment (oshoval) [inline .github/workflows/kubevirtci-tag-bumper.yaml:5]: please change to 3M, worth to change the comment to say it is 3M

**Comment by ashokpariya0** [comment]
Okay thanks, Yes, I agree with you guys; there's no need to add such checks, We'll see if anything fails.

**Comment by RamLavi** [comment]
> > We can add another workflow to bump to the maximum KUBEVIRT_PROVIDER (with format of k8s-x.y). wdyt?
> 
> it is problematic, because sometimes latest kubevirtci is more advance than the one we are working on, since the upstream CI team adds it before its needed
> 
> for sure we can discuss it on follow ups with Brian and Daniel
> 
> thanks

Then I'll try to add a step that uses the tag chosen to figure out the latest provider, and use that

**Comment by oshoval** [comment]
> > > We can add another workflow to bump to the maximum KUBEVIRT_PROVIDER (with format of k8s-x.y). wdyt?
> > 
> > 
> > it is problematic, because sometimes latest kubevirtci is more advance than the one we are working on, since the upstream CI team adds it before its needed
> > for sure we can discuss it on follow ups with Brian and Daniel
> > thanks
> 
> Then I'll try to add a step that uses the tag chosen to figure out the latest provider, and use that

it isnt good, because latest might be newer than the one we use, because CI team adds the cutting edge, yet used version

**Review by oshoval — APPROVED** [review]

File: .github/workflows/kubevirtci-tag-bumper.yaml (hunk 2 above)
Comment (RamLavi) [reply .github/workflows/kubevirtci-tag-bumper.yaml]: I would prefer to keep as is

File: .github/workflows/kubevirtci-tag-bumper.yaml (hunk 3 above)
Comment (RamLavi) [reply .github/workflows/kubevirtci-tag-bumper.yaml]: DONE

**Comment by RamLavi** [comment]
@oshoval added a step to also bump kubevirt-provider. PTAL

File: hack/bump-kubevirt-provider.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash -e
+
+[[ -z "$1" ]] && echo "error: missing required tag argument" >&2 && exit 1
+TAG="$1"
+
+LATEST_K8S=$(curl -s "https://api.github.com/repos/kubevirt/kubevirtci/contents/cluster-up/cluster?ref=${TAG}" \
+  | grep '"name": "k8s-[0-9]\+\.[0-9]\+"' \
+  | sed -E 's/.*"k8s-([0-9]+\.[0-9]+)".*/\1/' \
+  | sort -V \
+  | tail -n1)
```
Comment (oshoval) [inline hack/bump-kubevirt-provider.sh:10]: Thanks
didn't look on all deeply yet but:

we cant take `tail -n1`
even if we wanted be on the bleeding edge (i.e k8s version higher than the one that is recommended at the very moment)
we can't, because when we freeze and create a branch, the branch need to use the recommended one

i.e release-0.97 uses k8s version X, not X+1 even if kubevirtci already has it

what you can do is to count, and always take the 3rd starting from oldest
on most of the cases it will work, it would stop working if the providers will have suddenly different naming,
as happened in the past, where we had
k8s-x-centos9 and such, which might breaks this formula, but we can do it meanwhile, and be proactive according needs

Comment (RamLavi) [reply hack/bump-kubevirt-provider.sh:10]: ah I see. 
>what you can do is to count, and always take the 3rd starting from oldest

why third? I don't get the logic

Comment (RamLavi) [reply hack/bump-kubevirt-provider.sh:10]: Sorry I don't get it.. How do we know which version is "the one we officially need to support"? can't we just use that?

Comment (RamLavi) [reply hack/bump-kubevirt-provider.sh:10]: In the end we need to agree on the logic "which of the 3 will it choose"

Comment (oshoval) [reply hack/bump-kubevirt-provider.sh:10]: lets discuss offline and split to 2 PRs ?

Comment (oshoval) [reply hack/bump-kubevirt-provider.sh]: https://github.com/kubevirt/cluster-network-addons-operator/pull/2216#discussion_r2026834456

> Removing the KUBEVIRT_PROVIDER set. The nightly should use the default.

Food for thought: 
This can be an idea how to know the suggested provider to use,
we can suggest on kubevirtci, that it will know to use the latest by default, it will know the formula
and also will handle rare corner cases like we discussed

**Comment by oshoval** [comment]
/lgtm
would you able to trigger it to test it please?

unless we are already on latest

**Comment by RamLavi** [comment]
@oshoval once it's merged, I'll run it.
/approve
